### PR TITLE
syncer(dm): fix log error caused by "COMMIT" in QueryEvent

### PR DESCRIPTION
**Pull request by liumengya94**
### What problem does this PR solve?

Issue Number: ref #7525

### What is changed and how it works?
fix log error and save checkpoint for COMMIT QueryEvent

#### Tests 
 - Integration test

##### Will it cause performance regression or break compatibility?
No
##### Do you need to update user documentation, design documentation or monitoring documentation?
No
### Release note 

```release-note
None
```

### CODE REVIEW

File: dm/syncer/syncer.go [hunk 1]
```diff
@@ -2777,7 +2788,7 @@ func (s *Syncer) trackDDL(usedSchema string, trackInfo *ddlInfo, ec *eventContex
 
 func (s *Syncer) trackOriginDDL(ev *replication.QueryEvent, ec eventContext) (map[string]map[string]struct{}, error) {
 	originSQL := strings.TrimSpace(string(ev.Query))
-	if originSQL == "BEGIN" || originSQL == "" || utils.IsBuildInSkipDDL(originSQL) {
+	if originSQL == "BEGIN" || originSQL == "COMMIT" || originSQL == "" || utils.IsBuildInSkipDDL(originSQL) {
```
Comment (lance6716) [inline dm/syncer/syncer.go:2791]: seems because we have checked "COMMIT" in line 2275, this check is not needed.

File: dm/syncer/syncer.go [hunk 2]
```diff
@@ -2272,7 +2272,18 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
 			}
 		case *replication.QueryEvent:
 			originSQL = strings.TrimSpace(string(ev.Query))
-			err2 = s.ddlWorker.HandleQueryEvent(ev, ec, originSQL)
+			if originSQL == "COMMIT" {
```
Comment (lance6716) [inline dm/syncer/syncer.go:2275]: maybe use a `fallthrough` to run into the `case *replication.XIDEvent:` branch

File: dm/syncer/syncer.go [hunk 3]
```diff
@@ -2290,16 +2274,50 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
...
+					return true, nil
 				}
 			}
 
 			s.tctx.L().Debug("", zap.String("event", "XID"), zap.Stringer("last location", lastTxnEndLocation), log.WrapStringerField("location", endLocation))
 
 			job := newXIDJob(endLocation, startLocation, endLocation)
-			_, err2 = s.handleJobFunc(job)
+			_, err = s.handleJobFunc(job)
+			return needContinue, err
```
Comment (lance6716) [inline dm/syncer/syncer.go:2288]:
```suggestion
			return false, err
```

use explicit value for better understanding

File: dm/syncer/syncer.go [hunk 4]
```diff
@@ -2290,16 +2274,50 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
...
+			if sourceTable != nil && err2 == nil && s.cfg.EnableGTID {
+				if _, ok := affectedSourceTables[sourceTable.Schema]; !ok {
+					affectedSourceTables[sourceTable.Schema] = make(map[string]struct{})
+				}
+				affectedSourceTables[sourceTable.Schema][sourceTable.Name] = struct{}{}
+			}
+		case *replication.QueryEvent:
+			originSQL = strings.TrimSpace(string(ev.Query))
+			if originSQL == "COMMIT" {
+				needContinue, _ := funCommit()
```
Comment (lance6716) [inline dm/syncer/syncer.go:2307]: don't call `funCommit` twice

File: dm/tests/start_task/run.sh [hunk 5]
```diff
@@ -240,6 +240,69 @@ function run() {
...
+	done
+}
+
+function test_COMMIT_in_QueryEvent() {
+	echo "[$(date)] <<<<<< start test_COMMIT_in_QueryEvent >>>>>>"
+	cleanup_process
+	cleanup_data start_task
+	prepare_data_MyISAM
+
+	cp $cur/conf/source1.yaml $WORK_DIR/source1.yaml
```
Comment (lance6716) [inline dm/tests/start_task/run.sh:263]: no need to copy them if you won't change the file content

File: dm/syncer/syncer.go [hunk 6]
```diff
@@ -2290,16 +2275,47 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
...
 					if err != nil {
-						return terror.Annotatef(err, "shard group current location %s", shardingReSync.currLocation)
+						return false, terror.Annotatef(err, "shard group current location %s", shardingReSync.currLocation)
 					}
-					continue
+					return true, nil
 				}
 			}
 
 			s.tctx.L().Debug("", zap.String("event", "XID"), zap.Stringer("last location", lastTxnEndLocation), log.WrapStringerField("location", endLocation))
```
Comment (lance6716) [inline dm/syncer/syncer.go:2284]:
```suggestion
			s.tctx.L().Debug("", zap.String("event", "XID or COMMIT query event"), zap.Stringer("last location", lastTxnEndLocation), log.WrapStringerField("location", endLocation))
```

**Review by lance6716 — APPROVED** [review]

File: dm/syncer/syncer.go [hunk 7]
```diff
@@ -2256,24 +2256,9 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
...
-				if _, ok := affectedSourceTables[sourceTable.Schema]; !ok {
-					affectedSourceTables[sourceTable.Schema] = make(map[string]struct{})
-				}
-				affectedSourceTables[sourceTable.Schema][sourceTable.Name] = struct{}{}
-			}
-		case *replication.QueryEvent:
-			originSQL = strings.TrimSpace(string(ev.Query))
-			err2 = s.ddlWorker.HandleQueryEvent(ev, ec, originSQL)
-		case *replication.XIDEvent:
+		funCommit := func() (needContinue bool, err error) {
```
Comment (D3Hunter) [inline dm/syncer/syncer.go:2261]:
```suggestion
		funcCommit := func() (bool, error) {
```
this function doesn't use the named return var `needContinue`

File: dm/tests/start_task/run.sh [hunk 8]
```diff
@@ -240,6 +240,69 @@ function run() {
...
+	run_dm_ctl_with_retry $WORK_DIR "127.0.0.1:$MASTER_PORT" \
+		"query-status test" \
+		"\"result\": true" 2 \
+		"\"unit\": \"Sync\"" 1 \
+		"\"stage\": \"Running\"" 2
+
+	echo "check data"
+	check_sync_diff $WORK_DIR $cur/conf/diff_config.toml
+
+	check_log_not_contains $WORK_DIR/worker1/log/dm-worker.log "originSQL: COMMIT"
```
Comment (D3Hunter) [inline dm/tests/start_task/run.sh:303]: where does this log printed previously?

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

File: dm/tests/start_task/run.sh (hunk 8 above)
Comment (liumengya94) [reply dm/tests/start_task/run.sh:303]: in `trackOriginDDL`, "COMMIT" QueryEvent is identified as ddl, and print log in `parseOneStmt `https://github.com/pingcap/tiflow/blob/3f4a8a731c4cc7524ee69b621be2dd517170d776/dm/syncer/syncer.go#L2797
https://github.com/pingcap/tiflow/blob/3f4a8a731c4cc7524ee69b621be2dd517170d776/dm/syncer/ddl.go#L934
[2022/11/28 14:48:30.512 +08:00] [INFO] [ddl.go:934] ["parse ddl"] [task=test] [unit="binlog replication"] **[event=query] ["query event context"="{schema: , originSQL: COMMIT,** startLocation: position: (mysql-bin|000001.000001, 4407), gtid-set: e266a895-6e51-11ed-b110-0242ac110002:1-16, endLocation: position: (mysql-bin|000001.000001, 4476), gtid-set: e266a895-6e51-11ed-b110-0242ac110002:1-16, lastLocation: position: (mysql-bin|000001.000001, 4476), gtid-set: e266a895-6e51-11ed-b110-0242ac110002:1-16, re-sync: , needHandleDDLs: , trackInfos: }"]

**Review by D3Hunter — APPROVED** [review]

**Pull request merged**